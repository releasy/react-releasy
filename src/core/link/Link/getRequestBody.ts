import { RequestNode, UploadableMap, Variables } from 'relay-runtime';

const getRequestBodyWithUploadables = (
  request: RequestNode,
  variables: Variables,
  uploadables: UploadableMap,
) => {
  const formData = new FormData();
  formData.append('query', request.text);
  formData.append('variables', JSON.stringify(variables));

  Object.keys(uploadables).forEach((key) => {
    formData.append(key, uploadables[key]);
  });

  return formData;
};

const getRequestBodyWithoutUplodables = (request: RequestNode, variables: Variables): string => {
  return JSON.stringify({
    variables,
    query: request.text,
  });
};

const getRequestBody = (
  request: RequestNode,
  variables: Variables,
  uploadables?: UploadableMap,
): FormData | string => {
  if (uploadables) {
    return getRequestBodyWithUploadables(request, variables, uploadables);
  }

  return getRequestBodyWithoutUplodables(request, variables);
};

export default getRequestBody;
