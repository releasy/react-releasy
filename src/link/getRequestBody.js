// @flow
import type { RequestNode, UploadableMap, Variables } from 'relay-runtime';

const getRequestBodyWithUploadables = (
  request: RequestNode,
  variables: Variables,
  uploadables: UploadableMap,
) => {
  const formData = new FormData();
  formData.append('query', request.text);
  formData.append('variables', JSON.stringify(variables));

  uploadables.forEach((uploadable, key) => {
    // $FlowIgnore
    formData.append(`image${key}`, {
      name: uploadable.filename || `image${key}.jpg`,
      type: uploadable.mime || 'image/jpg',
      uri: uploadable.path,
    });
  });

  return formData;
};

const getRequestBodyWithoutUplodables = (request: RequestNode, variables: Variables): string => {
  return JSON.stringify({
    query: request.text,
    variables,
  });
};

const getRequestBody = (
  request: RequestNode,
  variables: Variables,
  uploadables: UploadableMap,
): FormData | string => {
  if (uploadables) {
    return getRequestBodyWithUploadables(request, variables, uploadables);
  }

  return getRequestBodyWithoutUplodables(request, variables);
};

export default getRequestBody;
