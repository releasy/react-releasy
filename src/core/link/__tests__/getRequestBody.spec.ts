import {
  createQueryRequest,
  createUploadables,
  createVariables,
} from '../../../../tests/fixtures/relay';

import getRequestBody from '../getRequestBody';

it('should create default request body', () => {
  const request = createQueryRequest();
  const variables = createVariables();

  const body = getRequestBody(request, variables);

  expect(body).toMatchSnapshot();
});

it('should create request body for uploadables', () => {
  const request = createQueryRequest();
  const variables = createVariables();
  const uploadables = createUploadables();

  const body = getRequestBody(request, variables, uploadables);

  expect(body).toMatchSnapshot();
});