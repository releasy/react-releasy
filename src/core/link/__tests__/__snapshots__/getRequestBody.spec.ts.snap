// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create default request body 1`] = `"{\\"variables\\":{},\\"query\\":\\"\\\\n    query MeQuery {\\\\n      me {\\\\n        id\\\\n        name\\\\n      }\\\\n    }\\\\n  \\"}"`;

exports[`should create request body for uploadables 1`] = `
FormData {
  Symbol(impl): FormDataImpl {
    "_entries": Array [
      Object {
        "name": "query",
        "value": "
    query MeQuery {
      me {
        id
        name
      }
    }
  ",
      },
      Object {
        "name": "variables",
        "value": "{}",
      },
      Object {
        "name": "test-0.txt",
        "value": FileImpl {
          "_buffer": Object {
            "data": Array [],
            "type": "Buffer",
          },
          "lastModified": 1538847573026,
          "name": "test-0.txt",
          "type": "",
          Symbol(wrapper): File {
            Symbol(impl): [Circular],
          },
        },
      },
    ],
    Symbol(wrapper): [Circular],
  },
}
`;
